### PR TITLE
Fix toStartOfWeek wrong result with DateTime64

### LINKED ISSUE #66442: toStartOfWeek returns invalid week when applied to 1970-01-01 as DateTime64
toStartOfWeek returns invalid week '2149-06-03' when applied to '1970-01-01' as DateTime64. I suspect there's some kind of overflow happening. It works fine on DataTime type.
https://fiddle.clickhouse.com/6f0d5d76-468c-46ef-8948-8f6490f24240

### PR DESCRIPTION
**Pull request by yariks5s**
Closes #66442

### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix for function `toStartOfWeek` which returned the wrong result with a small `DateTime64` value

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

File: src/Functions/DateTimeTransforms.h [hunk 1]
```diff
@@ -381,11 +381,13 @@ struct ToStartOfWeekImpl
 
     static UInt16 execute(Int64 t, UInt8 week_mode, const DateLUTImpl & time_zone)
     {
-        return time_zone.toFirstDayNumOfWeek(time_zone.toDayNum(t), week_mode);
+        const auto & res = time_zone.toFirstDayNumOfWeek(time_zone.toDayNum(t), week_mode);
+        return res >= 0 ? res : 0;
```
Comment (evillique) [inline src/Functions/DateTimeTransforms.h:385]:
```suggestion
        return std::max(res, 0);
```

File: src/Functions/DateTimeTransforms.h [hunk 2]
```diff
@@ -381,11 +381,13 @@ struct ToStartOfWeekImpl
 
     static UInt16 execute(Int64 t, UInt8 week_mode, const DateLUTImpl & time_zone)
     {
-        return time_zone.toFirstDayNumOfWeek(time_zone.toDayNum(t), week_mode);
+        const auto & res = time_zone.toFirstDayNumOfWeek(time_zone.toDayNum(t), week_mode);
```
Comment (evillique) [inline src/Functions/DateTimeTransforms.h:384]: Also, I don't think & is needed here

**Review by evillique — APPROVED** [review]

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit 34ca3128ed0de60e03a105c43dc0924541f4a2c1 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/67558/34ca3128ed0de60e03a105c43dc0924541f4a2c1/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Style check</td><td>Runs a set of checks to keep the code style clean. If some of tests failed, see the related log from the report</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/67558/34ca3128ed0de60e03a105c43dc0924541f4a2c1/style_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

**Comment by evillique** [comment]
From the logs, it seems this is not a missing include problem but rather std::max not having specialization for our types. One way to solve this is to perform static_cast on arguments or to just make it the way it was before :)

**Comment by clickgapai** [comment]
Hi @yariks5s @evillique — while reviewing this PR I found the following:

- **Bug (P1):** `ToStartOfWeekImpl::execute(Int32, ...)` missing `std::max(res, 0)` clamp — Date32 overflow persists — https://github.com/ClickHouse/ClickHouse/issues/104970

Happy to discuss — close anything that's wrong or already addressed.

**Pull request merged**